### PR TITLE
Rearrange Docker docs

### PR DESCRIPTION
**Pull request by davidcassany**
This commit rearranges the Docker image builds documentation
in order to match the criteria exposed in #323.

### CODE REVIEW

**Comment by schaefi** [comment]
Thanks David

@tomschr could you have a short look ?

File: doc/source/building/build_docker_container.rst [hunk 1]
```diff
@@ -1,190 +1,91 @@
...
-Container Configuration
-~~~~~~~~~~~~~~~~~~~~~~~
-
-A docker container includes a set of information also called metadata.
-The metadata information is provided within the KIWI XML description
-as part of the ``<containerconfig>`` section. The following parameters
-can be specified:
+The Docker configuration metadata is provided to KIWI as part of the
+:ref:`XML description file <decription_components>` using the
+``<containerconfig>`` section. The following parameters
```
Comment (tomschr) [inline doc/source/building/build_docker_container.rst]: As we talk about XML, we should better use the term "tag" or "element" instead of "section".

Comment (tomschr) [inline doc/source/building/build_docker_container.rst]: Same with "parameters". I guess, these are (XML) attributes, right? If yes, better use "attributes" here.

File: doc/source/building/build_docker_container.rst [hunk 2]
```diff
@@ -1,190 +1,91 @@
...
-* ``cmd``: Provides the default execution parameters of the container.
-* ``label``: Adds custom metadata to an image, is key-value pair.
+* ``subcommand``: Provides the default execution parameters of the container.
+  Equivalent of the `CMD` directive of a Docker file.
+* ``labels``: Adds custom metadata to an image, is key-value pair. Equivalent
+  of the `LABEL` directive of a Docker file.
 * ``expose``: Informs at which ports is the container listening at runtime.
-* ``env``: Sets an environment value, is key-value pair.
+  Equivalent of the `EXPOSE` directive of a Docker file.
+* ``environment``: Sets an environment values using key-value pairs.
```
Comment (tomschr) [inline doc/source/building/build_docker_container.rst]: Plural or singular? I assume plural here, in that case I would rephrase to:

    Set environment values using key-value pairs.

It would be helpful to add *how* a single key-value pair can be separated from another one. Use spaces, commas, or something different?

File: doc/source/building/build_docker_container.rst [hunk 3]
```diff
@@ -1,190 +1,91 @@
...
+  Equivalent of the `env` directive of a Docker file.
 * ``entrypoint``: Sets the executable that the container will run, it can
-  include parameters.
-* ``volume``: Creates a mountpoint with the given name and marks it to hold
-  external volumes from the host or from other contianers.
-* ``user``: Sets the user name or UID to be used when running `cmd` and
-  `entrypoint`.
+  include parameters. Equivalent of the `ENTRYPOINT` directive of a Docker
+  file..
+* ``volumes``: Creates mountpoints with the given name and marks it to hold
```
Comment (tomschr) [inline doc/source/building/build_docker_container.rst]: Create mountpoints...

File: doc/source/building/build_docker_container.rst [hunk 4]
```diff
@@ -1,190 +1,91 @@
...
+  external volumes from the host or from other contianers. Equivalent of the
+  `VOLUME` directive of a Docker file.
+* ``user``: Sets the user name or UID to be used when running `entrypoint` and
```
Comment (tomschr) [inline doc/source/building/build_docker_container.rst]: I would slightly rephrase it as:  "Sets the user name or user id (UID)..."

File: doc/source/building/build_docker_container.rst [hunk 5]
```diff
@@ -1,190 +1,91 @@
...
-
-* `ADD`
+  `entrypoint`. Equivalent of the `WORKDIR` directive of a Docker file.
 
-  In order to include a tarball and auto extract it in KIWI it can be done by
-  including the files in the :ref:`overlay tree <decription_components>`,
-  or by including an `<archive>` item inside the `<packages type="image">`
-  section. **ADD** also handles remote content, in KIWI this should be done
-  within the `config.sh` script.
+Other Docker file directives such as ``RUN`` and ``COPY`` or ``ADD``, can be
```
Comment (tomschr) [inline doc/source/building/build_docker_container.rst]: Correct with:

  ... ``RUN``, ``COPY`` or ``ADD``, ...

File: doc/source/building/build_docker_container.rst [hunk 6]
```diff
@@ -1,190 +1,91 @@
...
 
-  In order to execute commands during build time in KIWI it can be done by
-  extending the `config.sh` script. Note that the `config.sh` is executed
-  after all the packages have been installed and after the overlay tree has
-  been applied.
+1. Make sure you have checked out the example image descriptions,
+   see :ref:`example-descriptions`.
 
-* `COPY`
+2. Install the required toolchain for Docker images building.
```
Comment (tomschr) [inline doc/source/building/build_docker_container.rst]: IMHO, I would separate the step of including the repo from installing the packages like this:

```
1. Include the ``Virtualization`` repository to your list:

   .. code:: bash
      $ zypper ar -f http://download.opensuse.org/repositories/Virtualization:/containers/<DIST>

2. Install the :command:`umoci` and :command:`skopeo` tools:
  [...]
```

File: doc/source/building/build_docker_container.rst [hunk 7]
```diff
@@ -1,190 +1,91 @@
...
-is that the base image must be specified in the description file, it can
-be done using the **derived_from** optional attribute of ``<type>`` tag.
-The value of the **derived_from** attribute is the URI of the image;
-currently KIWI only supports references to local files (it expects a xz
-compressed tarball image) and any other URI type that is supported by
-the skopeo tool (for example DockerHub URIs as
-``derived_from="docker://opensuse:leap"``). The following example
-type specification shows how to specify a base image setup:
+   where the placeholder `<DIST>` is the preferred distribution. Then install
+   :command:`umoci` and :command:`skopeo` tools
```
Comment (tomschr) [inline doc/source/building/build_docker_container.rst]: See [previous comment](https://github.com/SUSE/kiwi/pull/363/files#r120322349).

**Comment by davidcassany** [comment]
@tomschr I think I have addressed all your comments, I have divided the list of configurable parameters between attributes and tags, I think this way is more clear. I did not want to document how to explicitly configure the Docker image, I assume that the users of this feature are already familiar with the Dockerfile syntax, so I just wanted to give an overview at a glance of the configuration possibilities, rather than explain them in detail.

**Comment by schaefi** [comment]
recompiled the docs and to me it looks great 👍

**Comment by tomschr** [comment]
@davidcassany 
>  I think I have addressed all your comments,

Great, thank you for your efforts!

>  I have divided the list of configurable parameters between attributes and tags, I think this way is more clear.

Yes, much better! 👍

> I did not want to document how to explicitly configure the Docker image, I assume that the users of this feature are already familiar with the Dockerfile syntax, so I just wanted to give an overview at a glance of the configuration possibilities, rather than explain them in detail.

Yes, totally agree. If you like and know a good link (maybe upstream Docker docu?) you can reference it and delegate it to to remote documentation. No need to copy existing texts.

**Comment by tomschr** [comment]
@davidcassany, @schaefi after I've built the doc myself, I have a final question. 😉

How should we display commands that require root privileges?

In most cases, we use `$` as a prompt. IMHO it could be misleading. Although the experienced user would probably know how to deal with it, nonetheless we should be clear about that. I see the following solutions to it:

1. Use `#` to indicate the root prompt.
2. Use `sudo`.
3. Leave it as is and assume all users are experienced

This is mostly important for the `zypper` and `docker` commands, depending on how the user configured its system. Regardless what we use, we should be consistent. 😉

What do you think?

**Comment by schaefi** [comment]
I prefix any command with a '$' and for those which require sudo i write '$ sudo'. I'm not sure if this follows the common practice but I did that all over the documentation

**Comment by tomschr** [comment]
> I prefix any command with a '$' and for those which require sudo i write '$ sudo'.
> I'm not sure if this follows the common practice but I did that all over the documentation.

The only "common practice" is to be consistent. 😉 It doesn't matter if it is with `sudo` or with `#` as long as the reader knows what to do and what to expect.

So no need to change anything here. I just stumbled upon it. Sorry for the noise.
